### ISSUE TITLE
Add typehints to the Python speechd API

### ISSUE THREAD

**Issue by RastislavKish** [brailcom/speechd#976]
**Is your feature request related to a problem? Please describe.**

Currently, the speech dispatcher Python client does not include typehints. This prevents static analysis tools such as mypy from verifying correct usage of the library from outside, as well as the internal integrity.

**Describe the solution you'd like**

Add typehints to all external/internal methods and structures.

**Describe alternatives you've considered**

The module can be excluded from statical analysis by the developer, however adding the types is the preferred solution.

**Additional context**

I can take this on, since it relates to a project I'm currently working on. Just wanted to check with the project whether we're of the same mind, before starting the actual work. :)

**Comment by sthibaul** [comment]
> I can take this on, since it relates to a project I'm currently working on

Yes, please. Type hints are good for people reading the code too, anyway :)

**Comment by RastislavKish** [comment]
I'm going through the code, adding typehints wherever necessary. One problem which jumped out on me is the connection arguments parsing in the SSIPClient class. This is currently done by a quite hacky approach using dictionaries, which are an unfortunate solution since they don't carry any consistency guarantees, are hard to read and typecheckers can't see into them. In the __init__ method, the communication_method argument, for example, is misnamed as method at one place, this part of the code likely doesn't work in practice and it's something the typechecker should be able to catch easily, except right now it can't.

If you'd agree, I would rewrite the entire parsing logic in a more object-style manner, so we don't need to pass weird dictionaries across methods. The API would stay unchanged.

There is also a codeblock marked for removal in 0.8. Shall I get rid of its logic, or replicate it?

**Comment by sthibaul** [comment]
> If you'd agree, I would rewrite the entire parsing logic in a more object-style manner, so we don't need to pass weird dictionaries across methods. The API would stay unchanged

Yes, that would be fine.

> There is also a codeblock marked for removal in 0.8. Shall I get rid of its logic, or replicate it?

Apparently some software is still using them.

I have now reported it on https://gitlab.com/smc/ibus-braille/-/issues/4 https://github.com/brailcom/speechd-el/issues/8
I haven't finished checking software using the deprecated arguments of `SSIPClient`

**Comment by sthibaul** [comment]
>  I haven't finished checking software using the deprecated arguments of SSIPClient

I didn't find any in Debian that uses them, so that part can probably go.

**Comment by RastislavKish** [comment]
Nice, thanks! I will refactor it out then.

Also another thing popped up during statical analysis. Do we have a minimum supported python version?

```python
try:
    import threading
except:
    import dummy_threading as threading
```

dummy_threading no longer exists as of Python 3.9, which is 4 years old, and I'd guess it had probably been deprecated for some time at that point.

**Comment by sthibaul** [comment]
> Do we have a minimum supported python version?

I don't know.
